src/register/pmpaddrx.rs

```diff
@@ -1,68 +1,100 @@
 use crate::register::pmpcfgx::Mode;
 use bit_field::BitField;
-use core::num::NonZeroUsize;
+use core::num::NonZeroU64;
 
 #[derive(Copy, Clone, Debug)]
 pub struct PmpAddr {
     bits: usize,
 }
 
+pub type Size = u64;
+pub type NonZeroSize = NonZeroU64;
+pub type Addr = u64;
+
 impl PmpAddr {
     #[inline]
-    pub fn decode(&self, mode: Mode) -> (usize, Option<NonZeroUsize>) {
+    pub fn decode(&self, mode: Mode) -> (Option<Addr>, Option<NonZeroSize>) {
+        let big_bits: Addr = self.bits as Addr;
         match mode {
-            Mode::OFF => (self.bits, None),
-            Mode::TOR => (self.bits << 2, None),
-            Mode::NA4 => (self.bits << 2, None),
+            Mode::OFF => (None, None),
+            Mode::TOR => (Some(big_bits << 2), None),
+            Mode::NA4 => (Some(big_bits << 2), Some(4.try_into().unwrap())),
             Mode::NAPOT => {
                 let (addr, size) = Self::decode_napot(self.bits);
-                (addr, Some(size.try_into().unwrap()))
+                (Some(addr), Some(size.try_into().unwrap()))
             }
         }
     }
 
     #[inline]
-    pub fn encode(&mut self, mode: Mode, addr: usize, size: Option<NonZeroUsize>) {
+    pub fn encode(&mut self, mode: Mode, addr: Addr, size: Option<NonZeroSize>) -> Result<(), ()> {
```

Comment:
Neither the Err nor Ok case carry information, why not leave this without any return type?

Reply:
It is to represent whether or not the function succeeded.  If `Ok(())` is returned, it means the internal bit field was written, if `Error(())` was returned, the internal bit field was not written.  Hopefully the caller of this function can use this information to fail gracefully and/or handle the error, rather than immediately panic.